File: src/pretty-print.ts
import {
  Program,
  Function,
  Stmt,
  Expr,
  Constant,
  UnaryExpr,
  BinaryExpr,
  BlockItem,
  Declaration,
  Assignment,
  Var,
  IfStmt,
  Conditional,
} from "./ast";
const INDENT_INCREMENT = 2;

function printLine(line: string, indentLevel: number): string {
  return " ".repeat(indentLevel) + line + "\n";
}

function printConstant(c: Constant): string {
  switch (c.kind) {
    case "numeric-const": {
      return `${c.value}`;
    }
    case "string-const": {
      return `${c.value}`;
    }
    default: {
      const _check: never = c;
      return _check;
    }
  }
}

function printBinary(binOp: BinaryExpr): string {
  return `${binOp.operator}(${printExpr(binOp.lhs)}, ${printExpr(binOp.rhs)})`;
}

function printUnary(u: UnaryExpr): string {
  return `${u.op}(${printExpr(u.expr)})`;
}

function printAssignment(assign: Assignment): string {
  // TODO parens?
  return `=(${printExpr(assign.lhs)}, ${printExpr(assign.rhs)})`;
}

function printVar(v: Var): string {
  return v.name;
}

function printConditional(exp: Conditional): string {
  return `Conditional(${printExpr(exp.cond)}, ${printExpr(
    exp.trueExpr
  )}, ${printExpr(exp.falseExpr)})`;
}

function printExpr(exp: Expr): string {
  switch (exp.kind) {
    case "numeric-const":
    case "string-const": {
      return printConstant(exp);
    }
    case "complement":
    case "unary-minus":
    case "logical-not": {
      return printUnary(exp);
    }
    case "binary-expr": {
      return printBinary(exp);
    }
    case "assignment": {
      return printAssignment(exp);
    }
    case "var": {
      return printVar(exp);
    }
    case "conditional": {
      return printConditional(exp);
    }
    default: {
      const _checker: never = exp;
      return _checker;
    }
  }
}

function printIfStmt(stmt: IfStmt, indentLevel: number): string {
  let res = printLine(`If(${printExpr(stmt.cond)}) {`, indentLevel);
  const childIndent = indentLevel + INDENT_INCREMENT;
  res += printStmt(stmt.thenStmt, childIndent);
  if (stmt.elseStmt) {
    res += printLine("} Else {", indentLevel);
    res += printStmt(stmt.elseStmt, childIndent);
    res += printLine("}", indentLevel);
  } else {
    res += printLine("}", indentLevel);
  }
  return res;
}

function printStmt(stmt: Stmt, indentLevel: number): string {
  switch (stmt.kind) {
    case "return-stmt": {
      const exp = printExpr(stmt.expr);
      return printLine(`return ${exp};`, indentLevel);
    }
    case "expr-stmt": {
      const exp = printExpr(stmt.expr);
      return printLine(`${exp};`, indentLevel);
    }

    case "null-stmt": {
      return printLine(`;`, indentLevel);
    }

    case "if-stmt": {
      return printIfStmt(stmt, indentLevel);
    }
  }
}

function printDeclaration(decl: Declaration, indentLevel: number): string {
  const init = decl?.init ? `,${printExpr(decl.init)}` : "";
  return printLine(`Declaration (${decl.name}${init});`, indentLevel);
}

function printBlock(block: BlockItem, indentLevel: number): string {
  switch (block.kind) {
    case "declaration":
      return printDeclaration(block, indentLevel);
    case "return-stmt":
    case "expr-stmt":
    case "null-stmt":
    case "if-stmt":
      return printStmt(block, indentLevel);
  }
}

function printFunction(func: Function, indentLevel: number): string {
  const body = func.body
    .map((block) => printBlock(block, indentLevel + INDENT_INCREMENT))
    .join("");
  let ret = printLine(`Function ${func.name}() {`, indentLevel);
  ret += body;
  ret += printLine("}", indentLevel);
  return ret;
}

function printProgram(program: Program, indentLevel: number): string {
  const f = printFunction(
    program.function_definition,
    indentLevel + INDENT_INCREMENT
  );
  return `${printLine("Program (", indentLevel)}${f}${printLine(
    ")",
    indentLevel
  )}`;
}

export function prettyPrint(ast: Program): string {
  return printProgram(ast, 0);
}

File: rust_src/pretty_print.rs
use crate::{ast, shared_types::Identifier};

const INDENT_INCREMENT: usize = 2;

fn print_line(line: &str, indent: usize) -> String {
  " ".repeat(indent).to_string() + line + "\n"
}

fn print_conditional(
  cond: &ast::ExprRef,
  true_expr: &ast::ExprRef,
  false_expr: &ast::ExprRef,
) -> String {
  format!(
    "Conditional({}, {}, {})",
    print_expr(cond.as_ref()),
    print_expr(true_expr.as_ref()),
    print_expr(false_expr.as_ref())
  )
}

fn print_assignment(left: &ast::ExprRef, right: &ast::ExprRef) -> String {
  format!(
    "=({}, {})",
    print_expr(left.as_ref()),
    print_expr(right.as_ref())
  )
}

fn print_var(name: &Identifier) -> String {
  format!("${name}")
}

fn print_binary(op: &ast::BinaryOp, left: &ast::ExprRef, right: &ast::ExprRef) -> String {
  format!(
    "{op:?}({}, {})",
    print_expr(left.as_ref()),
    print_expr(right.as_ref())
  )
}

fn print_unary(op: &ast::UnaryOperator, expr: &ast::ExprRef) -> String {
  format!("{op:?}({})", print_expr(expr.as_ref()))
}

fn print_expr(expr: &ast::Expr) -> String {
  match expr {
    ast::Expr::IntConstant(val) => format!("${val}"),
    ast::Expr::UnaryExpr(op, expr) => print_unary(op, expr),
    ast::Expr::BinaryExpr(binary_op, left, right) => print_binary(binary_op, left, right),
    ast::Expr::Var(name) => print_var(name),
    ast::Expr::Assignment(left, right) => print_assignment(left, right),
    ast::Expr::Conditional {
      cond,
      true_expr,
      false_expr,
    } => print_conditional(cond, true_expr, false_expr),
  }
}

fn print_declaration(id: &Identifier, init: &Option<ast::Expr>, indent: usize) -> String {
  let init_str = init
    .as_ref()
    .map_or("".to_string(), |e| format!(", {}", print_expr(e)));
  print_line(&format!("Declaration({}{});", id, init_str), indent)
}

fn print_if(
  cond: &ast::Expr,
  true_stmt: &Box<ast::Stmt>,
  false_stmt: &Option<Box<ast::Stmt>>,
  indent: usize,
) -> String {
  let mut res = print_line(&format!("If({}) {{", print_expr(cond)), indent);
  let child_indent = indent + INDENT_INCREMENT;
  res += &print_stmt(true_stmt, child_indent);
  if let Some(false_stmt) = false_stmt {
    res += &print_line("} Else {", indent);
    res += &print_stmt(false_stmt.as_ref(), child_indent);
  }
  res += &print_line("}", indent);
  res
}

fn print_stmt(stmt: &ast::Stmt, indent: usize) -> String {
  match stmt {
    ast::Stmt::Return(expr) => {
      let expr = print_expr(expr);
      print_line(&format!("return {};", expr), indent)
    }
    ast::Stmt::Expr(expr) => {
      let expr = print_expr(expr);
      print_line(&format!("{};", expr), indent)
    }
    ast::Stmt::Null => print_line(";", indent),
    ast::Stmt::If {
      cond,
      true_stmt,
      false_stmt,
    } => print_if(cond, true_stmt, false_stmt, indent),
  }
}

fn print_block(block: &ast::BlockItem, indent: usize) -> String {
  match &block {
    ast::BlockItem::Declaration(id, init) => print_declaration(id, init, indent),
    ast::BlockItem::Stmt(stmt) => print_stmt(&stmt, indent),
  }
}

fn print_function(func: &ast::Function, indent: usize) -> String {
  let body = func
    .body
    .iter()
    .map(|block| print_block(block, indent + INDENT_INCREMENT))
    .collect::<Vec<String>>()
    .join("");

  let mut ret = print_line(&format!("Function {}() {{", func.name), indent);
  ret += &body;
  ret += &print_line("}", indent);
  ret
}

fn print_program(prog: &ast::Program, indent: usize) -> String {
  let f = print_function(&prog.function, indent + INDENT_INCREMENT);
  format!(
    "{}{}{}",
    print_line("Program(", indent),
    f,
    print_line(")", indent)
  )
}

/// Pretty-print the given ast::Program to a String
pub fn pretty_print(prog: &ast::Program) -> String {
  print_program(prog, 0)
}

#[cfg(test)]
mod tests {
  use crate::{error::Error, lex, parse, source_files::SourceFile};

  use super::*;
  use pretty_assertions::assert_eq;

  fn run_pretty_printer(code: &str) -> Result<String, Error> {
    let source = SourceFile::new(code.to_string(), "bogus.c".to_string());
    let tokens = lex::lex(&source)?;
    let prog = parse::parse(&tokens, &source)?;
    Ok(pretty_print(&prog))
  }
  #[test]
  // This is a basic pretty printer test. The tests for the parser use the pretty
  // printer and also lock down output. So we don't duplicate those here.
  fn pretty_print_basic() {
    let code = r#"
int main(void) {
  int x;
  int y = 2;
  return y + 2;
}"#;
    assert_eq!(
      run_pretty_printer(code).unwrap(),
      r#"Program(
  Function main() {
    Declaration(x);
    Declaration(y, $2);
    return Plus($y, $2);
  }
)
"#
    );
  }
}
